src/crimson/common/logclient.cc

```diff
@@ -0,0 +1,371 @@
+      do_log(prio, s);
+  }
+}
+
+void LogChannel::do_log(clog_type prio, const std::string& s)
+{
+  if (CLOG_ERROR == prio) {
+    ldout(cct,-1) << "log " << prio << " : " << s << dendl;
+  } else {
+    ldout(cct,0) << "log " << prio << " : " << s << dendl;
+  }
+  LogEntry e;
+  e.stamp = ceph_clock_now();
+  // seq and who should be set for syslog/graylog/log_to_mon
+  e.addrs = parent->get_myaddrs();
+  e.name = parent->get_myname();
+  e.rank = parent->get_myrank();
+  e.prio = prio;
+  e.msg = s;
+  e.channel = get_log_channel();
+
+  // log to monitor?
+  if (log_to_monitors) {
+    e.seq = parent->queue(e);
+  } else {
+    e.seq = parent->get_next_seq();
+  }
+
+  // log to syslog?
+  if (do_log_to_syslog()) {
+    ldout(cct,0) << __func__ << " log to syslog"  << dendl;
+    e.log_to_syslog(get_log_prio(), get_syslog_facility());
+  }
+
+  // log to graylog?
+  if (do_log_to_graylog()) {
+    ldout(cct,0) << __func__ << " log to graylog"  << dendl;
+    graylog->log_log_entry(&e);
+  }
+}
+
+ceph::ref_t<Message> LogClient::get_mon_log_message(bool flush)
+{
+  if (flush) {
+    if (log_queue.empty())
+      return nullptr;
+    // reset session
+    last_log_sent = log_queue.front().seq;
+  }
+  return _get_mon_log_message();
+}
+
+bool LogClient::are_pending()
+{
+  return last_log > last_log_sent;
+}
+
+ceph::ref_t<Message> LogClient::_get_mon_log_message()
+{
+  if (log_queue.empty())
+    return {};
+
+  // only send entries that haven't been sent yet during this mon
+  // session!  monclient needs to call reset_session() on mon session
+  // reset for this to work right.
+
+  if (last_log_sent == last_log)
+    return {};
+
+  // limit entries per message
+  unsigned num_unsent = last_log - last_log_sent;
+  unsigned num_send;
+  if (cct->_conf->mon_client_max_log_entries_per_message > 0)
+    num_send = std::min(num_unsent, (unsigned)cct->_conf->mon_client_max_log_entries_per_message);
+  else
+    num_send = num_unsent;
+
+  ldout(cct,10) << " log_queue is " << log_queue.size() << " last_log " << last_log << " sent " << last_log_sent
+		<< " num " << log_queue.size()
+		<< " unsent " << num_unsent
+		<< " sending " << num_send << dendl;
+  ceph_assert(num_unsent <= log_queue.size());
+  std::deque<LogEntry>::iterator p = log_queue.begin();
+  std::deque<LogEntry> o;
+  while (p->seq <= last_log_sent) {
+    ++p;
+    ceph_assert(p != log_queue.end());
+  }
+  while (num_send--) {
+    ceph_assert(p != log_queue.end());
+    o.push_back(*p);
+    last_log_sent = p->seq;
+    ldout(cct,10) << " will send " << *p << dendl;
+    ++p;
+  }
+  
+  return ceph::make_message<MLog>(m_fsid,
+				  std::move(o));
+}
+
+void LogClient::_send_to_mon()
+{
+  ceph_assert(is_mon);
+  ceph_assert(messenger->get_myname().is_mon());
+  ldout(cct,10) << __func__ << " log to self" << dendl;
+  auto log = _get_mon_log_message();
+  //TODO: implement crimson counterpart(WIP)
+  //messenger->get_loopback_connection()->send_message2(std::move(log));
```

Comment:
I think it's unnecessary. This is really for in-minotor-living-`LogClient`-instance as `ceph_assert(is_mon)` is already there.